File: atmibroker-xatmi/src/test/cpp/TestTPRecv.cxx
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2008, Red Hat, Inc., and others contributors as indicated
 * by the @authors tag. All rights reserved.
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
#include <cppunit/extensions/HelperMacros.h>

#include "BaseServerTest.h"

#include "xatmi.h"

#include "TestTPRecv.h"

extern void testtprecv_service(TPSVCINFO *svcinfo);

void TestTPRecv::setUp() {
	sendbuf = NULL;
	rcvbuf = NULL;

	// Setup server
	BaseServerTest::setUp();

	// Do local work
	cd = -1;
	sendlen = strlen("hello") + 1;
	CPPUNIT_ASSERT((sendbuf = (char *) tpalloc((char*) "X_OCTET", NULL, sendlen)) != NULL);
	CPPUNIT_ASSERT((rcvbuf = (char *) tpalloc((char*) "X_OCTET", NULL, sendlen)) != NULL);
	strcpy(sendbuf, "hello");
	CPPUNIT_ASSERT(tperrno == 0);
	int toCheck = tpadvertise((char*) "TestTPRecv", testtprecv_service);
	CPPUNIT_ASSERT(tperrno == 0);
	CPPUNIT_ASSERT(toCheck != -1);
}

void TestTPRecv::tearDown() {
	// Do local work
	if (cd != -1) {
		::tpdiscon(cd);
	}
	::tpfree(sendbuf);
	::tpfree(rcvbuf);
	int toCheck = tpunadvertise((char*) "TestTPRecv");
	CPPUNIT_ASSERT(tperrno == 0);
	CPPUNIT_ASSERT(toCheck != -1);

	// Clean up server
	BaseServerTest::tearDown();
}

void TestTPRecv::test_tprecv_sendonly() {
	cd = ::tpconnect((char*) "TestTPRecv", sendbuf, sendlen, TPSENDONLY);
	long revent = 0;
	int result = ::tprecv(cd, &rcvbuf, &rcvlen, 0, &revent);
	CPPUNIT_ASSERT(tperrno== TPEPROTO);
	CPPUNIT_ASSERT(result == -1);
}

void testtprecv_service(TPSVCINFO *svcinfo) {
}

File: integration-tests/src/test/java/org/jboss/blacktie/jatmibroker/xatmi/TestTPRecv.java
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2008, Red Hat, Inc., and others contributors as indicated
 * by the @authors tag. All rights reserved.
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
package org.jboss.blacktie.jatmibroker.xatmi;

import junit.framework.TestCase;

import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.jboss.blacktie.jatmibroker.RunServer;
import org.jboss.blacktie.jatmibroker.core.conf.ConfigurationException;

public class TestTPRecv extends TestCase {
	private static final Logger log = LogManager.getLogger(TestTPRecv.class);
	private RunServer server = new RunServer();
	private Connection connection;
	private int sendlen;
	private Buffer sendbuf;
	private Session cd;

	public void setUp() throws ConnectionException, ConfigurationException {
		server.serverinit();
		server.tpadvertiseTestTPRecv();

		ConnectionFactory connectionFactory = ConnectionFactory
				.getConnectionFactory();
		connection = connectionFactory.getConnection();

		sendlen = "recv".length() + 1;
		sendbuf = connection.tpalloc("X_OCTET", null);
	}

	public void tearDown() throws ConnectionException, ConfigurationException {
		if (cd != null) {
			cd.tpdiscon();
		}
		connection.close();
		server.serverdone();
	}

	public void test_tprecv_sendonly() throws ConnectionException {
		log.info("test_tprecv_sendonly");
		cd = connection.tpconnect(server.getServiceNameTestTPRecv(), sendbuf,
				sendlen, Connection.TPSENDONLY);
		try {
			cd.tprecv(0);
			fail("expected proto error");
		} catch (ConnectionException e) {
			assertTrue((e.getEvent() == Connection.TPEV_SVCERR)
					|| (e.getTperrno() == Connection.TPEPROTO));
		}
	}
}
